magefile.go

```diff
@@ -1154,15 +1216,319 @@ func packageAgent(platforms []string, packagingFn func()) {
+	return bi, nil
+}
+
+func mapManifestPlatformToAgentPlatform(manifestPltf string) (string, bool) {
+	mappings := map[string]string{
+		"linux-x86_64":   "linux-amd64",
+		"windows-x86_64": "windows-amd64",
+		"darwin-x86_64":  "darwin-amd64",
+		"darwin-aarch64": "darwin-arm64",
+		"linux/x86_64":   "linux/amd64",
+		"windows/x86_64": "windows/amd64",
+		"darwin/x86_64":  "darwin/amd64",
+		"darwin/aarch64": "darwin/arm64",
+	}
+
+	mappedPltf, found := mappings[manifestPltf]
+	if !found {
+		// default to the manifest platform if no mapping is found
+		mappedPltf = manifestPltf
+	}
+
+	return mappedPltf, found
+
+}
+
+func filterPackagesByPlatform(pkgs map[string]tools.Package) map[string]tools.Package {
+	if mg.Verbose() {
+		log.Printf("unfiltered packages: %v", pkgs)
+	}
+	platforms := devtools.Platforms.Names()
+	filteredPackages := map[string]tools.Package{}
+	for pkgName, pkgDesc := range pkgs {
+		if mg.Verbose() {
+			log.Printf("checking if %s:%v should be included", pkgName, pkgDesc)
+		}
+		for _, pkgOS := range pkgDesc.Os {
+			platformString, _ := mapManifestPlatformToAgentPlatform(fmt.Sprintf("%s/%s", pkgOS, pkgDesc.Architecture))
+			if slices.Contains(platforms, platformString) {
+				if mg.Verbose() {
+					log.Printf("platforms include %s", platformString)
+				}
+				filteredPackages[pkgName] = pkgDesc
+				break
+			}
+		}
+	}
+	if mg.Verbose() {
+		log.Printf("filtered packages: %v", filteredPackages)
+	}
+	return filteredPackages
+}
+
+func downloadDRAArtifacts(ctx context.Context, manifestUrl string, downloadDir string, projects ...string) (map[string]tools.Package, error) {
+
+	build, err := manifest.DownloadManifest(manifestUrl)
+	if err != nil {
+		return nil, fmt.Errorf("downloading manifest from %q: %w", manifestUrl, err)
+	}
+
+	// Create a dir with the buildID at <downloadDir>/<buildID>
+	draDownloadDir := filepath.Join(downloadDir, build.BuildID)
+	err = os.MkdirAll(draDownloadDir, 0o770)
+
+	if err != nil {
+		return nil, fmt.Errorf("creating %q directory: %w", err)
+	}
+
+	if mg.Verbose() {
+		log.Printf("url:%q, err %v", manifestUrl, err)
+	}
+
+	// sync access to the downloadedArtifacts map
+	mx := new(sync.Mutex)
+	downloadedArtifacts := map[string]tools.Package{}
+	errGrp, errCtx := errgroup.WithContext(ctx)
+
+	for _, projectName := range projects {
+		project, ok := build.Projects[projectName]
+		if !ok {
+			return nil, fmt.Errorf("project %q not found in manifest at %q", projectName, manifestUrl)
+		}
+
+		if mg.Verbose() {
+			log.Printf("build %q project %s packages: %+v", build.BuildID, projectName, project)
+		}
+		// filter down the packages to the platforms we are building/support
+		filteredPackages := filterPackagesByPlatform(project.Packages)
+		if mg.Verbose() {
+			log.Printf("packages to download: %v", filteredPackages)
+		}
+		for pkgName, pkgDesc := range filteredPackages {
+			downloadFunc := func(pkgName string, pkgDesc tools.Package) func() error {
+				return func() error {
```

Comment:
[Suggestion]
It'd be more straight forward, more clear, to just build a closure instead of a function which returns a function. At 1st glace it looks like you're already calling the function before passing it to `errGrp`

Reply:
this is a closure that returns a function() (which is what the errGrp expects)
The reason why this is done this way is to create a copy of the iterator variables so that the internal values are not tied to the `for` iterators. When we switch to Go 1.22 we can simplify this.